### DISCUSSION TITLE
Define expected type of body in API endpoint.

### DISCUSSION THREAD

**Discussion by MJDeligan** [nuxt/nuxt#16807] [Questions]
Hey,

Endpoints in the server directory have a typed return value. This is pretty handy and I was wondering if something similar can be done the other way around. Is it possible to define a typescript interface that the endpoint expects the body of requests to adhere to?

For example, say I have an enpoint `/api/message.post.ts`
```ts

export interface MessageCreate {
  subject: string;
  body: string;
}

export default defineEventHandler((event) => {
  const body = useBody<MessageCreate>(event);
  //... use body here
}
```

Is there a way to type this, so that I get type hints for the body on the frontend?

```ts
await $fetch(
  '/api/message',
  {
    method: 'POST',
    body: ??? // have type hint here for MessageCreate
  }
);
```

Right now the body seems to always have type `Record<string, any> | BodyInit`.
I couldn't find anything about this in the docs. Sorry if it's already covered.

**Comment by sevillaarvin** [comment]
I don't think this is available yet. For this to work, at least the baseURL should not be redefined. Perhaps this could be a feature request in the issues page?

**Comment by virgosoy** [comment]
Work in progress now
https://github.com/nuxt/nuxt/discussions/21881#discussioncomment-6334136
